File: SOLID/code-examples/chapter-12.py
# FILE: chapter-12.py
# Chapter 12 — SOLID in Practice — Python

from abc import ABC, abstractmethod

# SOLID Principles working together in an e-commerce system

# SRP - Single Responsibility
class Product:
    def __init__(self, product_id, name, price):
        self.id = product_id
        self.name = name
        self.price = price

class Order:
    def __init__(self, order_id):
        self.id = order_id
        self.products = []
    
    def add_product(self, product):
        self.products.append(product)
    
    def get_total(self):
        return sum(product.price for product in self.products)

# OCP & DIP - Payment strategies (open for extension)
class PaymentProcessor(ABC):
    @abstractmethod
    def process_payment(self, amount):
        pass

class CreditCardProcessor(PaymentProcessor):
    def process_payment(self, amount):
        print(f"💳 Processing ${amount} via Credit Card")
        return True  # Simulated success

class PayPalProcessor(PaymentProcessor):
    def process_payment(self, amount):
        print(f"🅿️ Processing ${amount} via PayPal")
        return True  # Simulated success

# ISP - Segregated interfaces
class Notifiable(ABC):
    @abstractmethod
    def send_notification(self, message):
        pass

class Trackable(ABC):
    @abstractmethod
    def track_order(self, order_id):
        pass

# LSP - Substitutable implementations
class EmailNotificationService(Notifiable):
    def send_notification(self, message):
        print(f"📧 Email: {message}")

class SMSNotificationService(Notifiable):
    def send_notification(self, message):
        print(f"📱 SMS: {message}")

# Main service coordinating everything
class OrderService:
    def __init__(self, payment_processor, notification_service):
        self.payment_processor = payment_processor
        self.notification_service = notification_service
    
    def process_order(self, order):
        print(f"🛒 Processing order {order.id}")
        
        # Process payment
        payment_success = self.payment_processor.process_payment(order.get_total())
        
        if payment_success:
            # Send confirmation
            self.notification_service.send_notification(
                f"Order {order.id} confirmed!"
            )
            return True
        
        return False

def main():
    print("🏢 SOLID Principles in Practice (Python)")
    print("========================================\n")
    
    # Create order
    order = Order("ORD-001")
    order.add_product(Product("P1", "Laptop", 999.99))
    order.add_product(Product("P2", "Mouse", 29.99))
    
    print(f"📦 Order created with total: ${order.get_total()}\n")
    
    # Process with different payment methods and notifications
    print("Processing with Credit Card + Email:")
    service1 = OrderService(
        CreditCardProcessor(),
        EmailNotificationService()
    )
    service1.process_order(order)
    
    print("\nProcessing with PayPal + SMS:")
    service2 = OrderService(
        PayPalProcessor(),
        SMSNotificationService()
    )
    service2.process_order(order)
    
    print("\n💡 SOLID Principles Applied:")
    print("   🎯 SRP: Each class has one responsibility")
    print("   🔓 OCP: Can add new payment/notification methods")
    print("   🔄 LSP: All implementations are substitutable")
    print("   🔀 ISP: Interfaces are focused and specific")
    print("   ⬇️ DIP: Depends on abstractions, not concretions")

if __name__ == "__main__":
    main()

File: SOLID/code-examples/chapter-12.cpp
// FILE: chapter-12.cpp
// Chapter 12 — SOLID in Practice — C++

#include <iostream>
#include <vector>
#include <memory>
#include <string>
using namespace std;

// SOLID Principles working together in an e-commerce system

// SRP - Single Responsibility
class Product {
private:
    string id, name;
    double price;
public:
    Product(const string& id, const string& name, double price) 
        : id(id), name(name), price(price) {}
    
    string getId() const { return id; }
    string getName() const { return name; }
    double getPrice() const { return price; }
};

class Order {
private:
    string id;
    vector<Product> products;
public:
    Order(const string& id) : id(id) {}
    
    void addProduct(const Product& product) {
        products.push_back(product);
    }
    
    double getTotal() const {
        double total = 0;
        for (const auto& product : products) {
            total += product.getPrice();
        }
        return total;
    }
    
    const vector<Product>& getProducts() const { return products; }
    string getId() const { return id; }
};

// OCP & DIP - Payment strategies (open for extension)
class PaymentProcessor {
public:
    virtual bool processPayment(double amount) = 0;
    virtual ~PaymentProcessor() = default;
};

class CreditCardProcessor : public PaymentProcessor {
public:
    bool processPayment(double amount) override {
        cout << "💳 Processing $" << amount << " via Credit Card" << endl;
        return true; // Simulated success
    }
};

class PayPalProcessor : public PaymentProcessor {
public:
    bool processPayment(double amount) override {
        cout << "🅿️ Processing $" << amount << " via PayPal" << endl;
        return true; // Simulated success
    }
};

// ISP - Segregated interfaces
class Notifiable {
public:
    virtual void sendNotification(const string& message) = 0;
    virtual ~Notifiable() = default;
};

class Trackable {
public:
    virtual void trackOrder(const string& orderId) = 0;
    virtual ~Trackable() = default;
};

// LSP - Substitutable implementations
class EmailNotificationService : public Notifiable {
public:
    void sendNotification(const string& message) override {
        cout << "📧 Email: " << message << endl;
    }
};

class SMSNotificationService : public Notifiable {
public:
    void sendNotification(const string& message) override {
        cout << "📱 SMS: " << message << endl;
    }
};

// Main service coordinating everything
class OrderService {
private:
    unique_ptr<PaymentProcessor> paymentProcessor;
    unique_ptr<Notifiable> notificationService;
    
public:
    OrderService(unique_ptr<PaymentProcessor> processor, 
                 unique_ptr<Notifiable> notifier)
        : paymentProcessor(move(processor)), notificationService(move(notifier)) {}
    
    bool processOrder(const Order& order) {
        cout << "🛒 Processing order " << order.getId() << endl;
        
        // Process payment
        bool paymentSuccess = paymentProcessor->processPayment(order.getTotal());
        
        if (paymentSuccess) {
            // Send confirmation
            notificationService->sendNotification(
                "Order " + order.getId() + " confirmed!"
            );
            return true;
        }
        
        return false;
    }
};

int main() {
    cout << "🏢 SOLID Principles in Practice (C++)" << endl;
    cout << "======================================" << endl << endl;
    
    // Create order
    Order order("ORD-001");
    order.addProduct(Product("P1", "Laptop", 999.99));
    order.addProduct(Product("P2", "Mouse", 29.99));
    
    cout << "📦 Order created with total: $" << order.getTotal() << endl << endl;
    
    // Process with different payment methods and notifications
    cout << "Processing with Credit Card + Email:" << endl;
    OrderService service1(
        make_unique<CreditCardProcessor>(),
        make_unique<EmailNotificationService>()
    );
    service1.processOrder(order);
    
    cout << endl << "Processing with PayPal + SMS:" << endl;
    OrderService service2(
        make_unique<PayPalProcessor>(),
        make_unique<SMSNotificationService>()
    );
    service2.processOrder(order);
    
    cout << endl << "💡 SOLID Principles Applied:" << endl;
    cout << "   🎯 SRP: Each class has one responsibility" << endl;
    cout << "   🔓 OCP: Can add new payment/notification methods" << endl;
    cout << "   🔄 LSP: All implementations are substitutable" << endl;
    cout << "   🔀 ISP: Interfaces are focused and specific" << endl;
    cout << "   ⬇️ DIP: Depends on abstractions, not concretions" << endl;
    
    return 0;
}